Mathlib/CategoryTheory/MorphismProperty/Factorization.lean

```diff
@@ -0,0 +1,203 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+
+/-!
+# The factorization axiom
+
+In this file, we introduce a type-class `HasFactorization W₁ W₂`, which, given
+two classes of morphisms `W₁` and `W₂` in a category `C`, asserts that any morphism
+in `C` can be factored as a morphism in `W₁` followed by a morphism in `W₂`. The data
+of such factorizations can be packaged in the type `FactorizationData W₁ W₂`.
+
+This shall be used in the formalization of model categories for which the CM5 axiom
+asserts that any morphism can be factored as a cofibration followed by a trivial
+fibration (or a trivial cofibration followed by a fibration).
+
+We also provide a structure `FunctorialFactorizationData W₁ W₂` which contains
+the data of a functorial factorization as above. With this design, when we
+formalize certain constructions (e.g. cylinder objects in model categories),
+we may first construct them using using `data : FactorizationData W₁ W₂`.
+Without duplication of code, it shall be possible to show these cylinders
+are functorial when a term `data : FunctorialFactorizationData W₁ W₂` is available,
+the existence of which is asserted in the type-class `HasFunctorialFactorization W₁ W₂`.
+
+-/
+
+namespace CategoryTheory
+
+namespace MorphismProperty
+
+variable {C : Type*} [Category C] (W₁ W₂ : MorphismProperty C)
+
+/-- Given two classes of morphisms `W₁` and `W₂` on a category `C`, this is
+the data of the factorization of a morphism `f : X ⟶ Y` as `i ≫ p` with
+`W₁ i` and `W₂ p`. -/
+structure MapFactorizationData {X Y : C} (f : X ⟶ Y) where
+  /-- the intermediate object in the factorization -/
+  Z : C
+  /-- the first morphism in the factorization -/
+  i : X ⟶ Z
+  /-- the second morphism in the factorization -/
+  p : Z ⟶ Y
+  fac : i ≫ p = f := by aesop_cat
+  hi : W₁ i
+  hp : W₂ p
+
+attribute [reassoc (attr := simp)] MapFactorizationData.fac
+
+/-- The data of a term in `MapFactorizationData W₁ W₂ f` for any morphism `f`. -/
+abbrev FactorizationData := ∀ {X Y : C} (f : X ⟶ Y), MapFactorizationData W₁ W₂ f
+
+/-- The factorization axiom for two classes of morphisms `W₁` and `W₂` in a category `C`. It
+asserts that any morphism can be factored as a morphism in `W₁` followed by a morphism
+in `W₂`. -/
+class HasFactorization : Prop where
+  nonempty_mapFactorizationData {X Y : C} (f : X ⟶ Y) : Nonempty (MapFactorizationData W₁ W₂ f)
```

Comment:
Maybe we should make `fun {X Y : C} (f : X ⟶ Y) => Nonempty (MapFactorizationData W₁ W₂ f)` into a `MorphismProperty`? It will be useful for say immersions and quasi-projective morphisms (on Hartshorne).

Looking at the material below, I think this definition is quite orthogonal to them and will not shorten any proof though. What do you think?

Reply:
Thanks for the suggestion. I have added `MorphismProperty.comp` and a lemme `comp_eq_top_iff`. It also seems to me that it is essentially independent of the rest of the code though.

Reply:
Yeah I was expecting the answer to be "the idea might have potential but perhaps in a separate PR" but I brought it up just in case.